pkg/gadgets/socket-collector/gadget.go

```diff
@@ -72,71 +75,121 @@ func (t *Trace) Operation(trace *gadgetv1alpha1.Trace,
+	// we only need to run the tracer with one valid PID per pod/namespace
+	visitedPods := make(map[string]bool)
+
+	// Keep also track of the number of different pods and the discarded
+	// containers to then check that we did not miss a pod
+	totalPods := 0
+	discardedContainers := []pb.ContainerDefinition{}
+
+	for _, container := range filteredContainers {
+		key := container.Podname + "/" + container.Namespace
+		if _, ok := visitedPods[key]; !ok {
+			totalPods++
+
+			if container.Pid == 0 {
+				discardedContainers = append(discardedContainers, container)
+
+				log.Warningf("Gadget %s: Discarding empty PID of Container %q Pod %q "+
+					"Namespace %q. Look for another Container in the same Pod",
+					gadgetName, container.Name, container.Podname, container.Namespace)
 
-		// All containers inside a pod share the same network namespace
-		// thus we can just take the first valid pid we found
-		for _, pid = range pidsInPod {
-			if pid != 0 {
-				break
+				continue
 			}
+
+			// The stored value does not matter, we are just keeping
+			// track of the visited Pods per Namespace
+			visitedPods[key] = true
+
+			log.Infof("Gadget %s: Using PID %d to retrieve network namespace of Pod %q in Namespace %q",
+				gadgetName, container.Pid, container.Podname, container.Namespace)
+
+			podSockets, err := tracer.RunCollector(container.Pid, container.Podname,
+				container.Namespace, node)
+			if err != nil {
+				return "", err
+			}
+
+			allSockets = append(allSockets, podSockets...)
 		}
+	}
 
-		log.Infof("Gadget %s: Using PID %d to retrieve network namespaces of Podname %q in Namespace %q",
-			trace.Spec.Gadget, pid, trace.Spec.Filter.Podname, trace.Spec.Filter.Namespace)
+	// It is not a problem if we discarded some containers
+	// as long as we visited all the requested pods
+	if len(visitedPods) != totalPods {
 
-		if pid == 0 {
-			gadgets.SetStatusError(trace, fmt.Sprintf("Couldn't find a valid PID for Podname %q in Namespace %q",
-				trace.Spec.Filter.Podname, trace.Spec.Filter.Namespace))
-			return
+		// This should never happen
+		if len(discardedContainers) == 0 {
+			log.Warningf("Gadget %s: We did not discarded a container but did not visited all the Pods? "+
+				"len(visitedPods) %d totalPods %d filteredContainers %+v",
+				gadgetName, len(visitedPods), totalPods, filteredContainers)
+
+			return "", fmt.Errorf("impossible to find valid PID for requested Pods")
 		}
 
-		t.Start(trace, pid)
-	default:
-		gadgets.SetStatusError(trace, fmt.Sprintf("Unknown operation %q", operation))
+		// Just improving the UX by giving a more detailed error description
+		missingPods := ""
+		for _, container := range discardedContainers {
+			key := container.Podname + "/" + container.Namespace
+			if _, ok := visitedPods[key]; !ok {
+				if missingPods != "" {
+					missingPods += ", "
+				}
+				missingPods += key
+			}
+		}
+
+		return "", fmt.Errorf("impossible to find valid PID for %s", missingPods)
```

Comment:
I don't think we should pay that much attention to containers without a PID. The PID of the containers is used in other places, so if there is a container without a PID it's because there's a problem somewhere else.

Reply:
Wait, if there is a container with pid == 0, does the whole gadget fail?

Reply:
In the initial PR of socket-collector we decided to make the whole gadget fail (Do not show any result), if we are not able to retrieve the information for even only one of the requested Pods (i.e. there is not at least one container with pid != 0 for a given Pod). 

We made this decision because if we show the information of the rest of the Pods, it could be misunderstood as if the involved Pod does not have any socket.

Reply:
I think we could just make this fail if even one container has pid==0. As I said below, it's extremely rare to have a container with pid == 0 and I think we should even forbid that situation in the gadget tracer manager.  @alban do you have any opinion here?

Reply:
Understood, I will make the gadget fail if there is a container with pid == 0. BTW, this change will simplify this PR.